### ISSUE TITLE
Please add an ability to use different build tools for --docker-build command. Make it extensible

### ISSUE THREAD

**Issue by andrew-kurochkin** [telepresenceio/telepresence#3215]
We have a need to be able to use different build engines for Docker containers for the newly added --docker-build command. 

As a solution Telepresence can create an extensible solution that allows customers to use any container build tool that fit requirements. It can be similar to "git remote extensions" where you can add your extension to a special folder where git would search for a binary with a particular name base on protocol prefix. For the Telepresence solution it could be something like --docker-build-buildx, where buildx would be considered as a builder tool. Than Telepresence would look into it's internal folder with extensions and call a corresponding command. Something like ./container-builders/telepresence-to-buildx that will accept build parameters. Looking into the code the only requirement for that tool would be to print docker container image id into stdout.

An alternative solution would be to embed different builders into the code.

**Comment by thallgren** [comment]
We currently call `docker` without a path, so if you start telepresence with a path that finds a bespoke binary by that name that you have provided, then that binary will be used (not just for building though, it will need to be an extension of the normal `docker` binary). Would that solve this problem?

**Comment by awnion** [comment]
> We currently call `docker` without a path, so if you start telepresence with a path that finds a bespoke binary by that name that you have provided, then that binary will be used (not just for building though, it will need to be an extension of the normal `docker` binary). Would that solve this problem?

As for me, in general, it's not a good idea to change the behavior of the executable owned by another party. Especially if it's something as ubiquitous as the `docker` executable.

1) Some users might have different versions of docker already installed: e.g. `docker-desktop` and native package manager's `docker`. One more executable may lead to extra confusion.
2) Some people address the software supply chain by checking if critical binaries like `docker` are genuine.
3) Also CI tools might ignore PATH's `docker` executable and either use `/usr/local/bin/docker` anyway or use unix socket API instead.

Instead, it could be easily supported the way `git` does it. `git` supports custom <schema>:// via `git-remote-<schema>` executable. 

For example, here's what I do in my fork of `pkg/client/docker/image.go`:

```go
// ...
import "net/url"
// ...
func BuildImage(ctx context.Context, context string, buildArgs []string) (string, error) {
	var dockerBuildCmd string
	var args []string

	parsedUrl, err := url.Parse(context)

	if err != nil {
		dockerBuildCmd = "docker"
		args = []string{"build", "--quiet"}
	} else {
		dockerBuildCmd = "telepresence-build-" + parsedUrl.Scheme
		args = []string{}
	}

	cmd := proc.StdCommand(ctx, dockerBuildCmd, append(append(args, buildArgs...), context)...)

// ...
```

**Comment by thallgren** [comment]
@awnion to clarify, I didn't mean that you should swap out the docker binary. What I meant was that, as  a workaround, you can do

```console
$ PATH=<path to directory containing a bespoke docker>:$PATH telepresence ...
```
This would not interfere with another already installed docker. It would just require that your bespoke docker strips that added path and calls docker again.

I realise that it's not a perfect solution, but it should work without any modification to telepresence.

**Comment by thallgren** [comment]
@awnion this `git-remote-<schema>` that you mention, can you please provide a link to some docs about that?

**Comment by awnion** [comment]
> @awnion to clarify, I didn't mean that you should swap out the docker binary. What I meant was that, as a workaround, you can do
> 
> ```
> $ PATH=<path to directory containing a bespoke docker>:$PATH telepresence ...
> ```
> 
> This would not interfere with another already installed docker. It would just require that your bespoke docker strips that added path and calls docker again.
> 
> I realise that it's not a perfect solution, but it should work without any modification to telepresence.

TLDR: Unfortunately, it won't work. 

Let's say I have the following dummy bash implementation of `docker`:

```bash
#!/bin/bash
docker $@
```

Which docker will it call? It will call itself. And I can't rely on `/usr/local/bin/docker` because it's not always the case. And changing $PATH on the fly sounds like a bad idea.

**Comment by awnion** [comment]
> @awnion this `git-remote-<schema>` that you mention, can you please provide a link to some docs about that?

Sure

https://git-scm.com/docs/gitremote-helpers

**Comment by thallgren** [comment]
Yes, your dummy won't work, because it doesn't strip the path segment added before calling telepresence. That must be part of the design. The whole idea builds on changing the path twice in a controlled manner. Not sure it's such a bad idea. It works, if done right.

**Comment by Futurizt** [comment]
> Yes, your dummy won't work, because it doesn't strip the path segment added before calling telepresence. That must be part of the design. The whole idea builds on changing the path twice in a controlled manner. Not sure it's such a bad idea. It works, if done right.

This breaks a lot of security assumptions instead of just adding 8 lines of code following best practices of Git design. Why we would (not) do it? Do I miss something here?

**Comment by thallgren** [comment]
> Why we would (not) do it? Do I miss something here?

@Futurizt, this is the answer I gave to Andrew K when he asked a similar question in Slack:

We will open up for a demand of extensions in general and questions like “if you provide an docker build extension point, then why don’t you also provide an extension point for X”.

I can see a potential for a large number of Xs here, some of which perhaps aren’t good extension candidates but will be implemented as such nevertheless. Each extension must be documented as a public API that we must consider in future versions. We will have to provide compatibility charts (i.e. what version of telepresence works with what version of X) and add regression tests that verifies those APIs.

In other words, I’m cautious to adding a new generic way to design and add potential future enhancements.

---

> This breaks a lot of security assumptions

Which assumptions are broken? The dual PATH modification is extremely isolated to one specific use-case. What problems does that introduce?

Nota bene, I still suggest this as a workaround until a more elaborate solution has been designed, but I feel that it's important to know what problems it causes since that might have an impact on the chosen final solution.

---

Regarding the git-remote extension:

This is an API where the extension must reply to "capabilities" and provide a list of `git remote` commands that it extends. Telepresence has a lot of subcommands that we could extend the same way, but `docker build` is a docker command. As such, it should be extended as a docker cli-plugin (but AFAICT, docker doesn't allow the `build` command to be extended). If we were to introduce an extension mechanism similar to `git-remote-<scheme>`, then we would dedicate that for a telepresence sub command. Also, git's way of extending `remote` using a URL scheme makes a lot of sense because it's intended for services like `git-remote-http`, `git-remote-https`, `git-remote-ftp`, and `git-remote-ftps`. All well-known schemes. Using a scheme to designate a build plugin doesn't make much sense IMO. 

In short, I'm not in favor of using a `git-remote-<scheme>` like extension mechanism to extend the docker build.

---

**Proposal**

Telepresence has a `config.yml` file where we can add new structured properties. The values of this config can also be set in the helm chart using the `client` property, so that they affect all clients that connect. We could add a way to configure the docker binary for builds here by providing a URL to download and install. This would allow an admin to configure the client globally so that it seamlessly downloads your plugin and then uses it for the docker build. Perhaps something like:

```yaml
docker:
  build:
    executable: https://your.download.site/{{ Runtime.GOOS }}/{{ Runtime.GOARCH }}/1.2.3/docker-build-gosh
```
with the assumption being that we pass everything after `docker build` to the downloaded binary and that it then performs the same tasks and produces the same output.

This would give us a common place to add other things that we might want to make configurable with respect to docker.

Caveat:
We're introducing a docker compose integration in a coming release. The compose spec may well contain services that are built on demand. I don't think we can swap the binary when that happens. Do you have a solution for this already?

**Comment by awnion** [comment]
> This is an API where the extension must reply to "capabilities" and provide a list of git remote commands that it extends. Telepresence has a lot of subcommands that we could extend the same way ...

It's not necessary to implement a full stdin-stdout protocol (as git does and we also do this in our `git-remote-gosh`). Yet the idea is that an end-user can implement their own executable with in some way a unique name (like `telepresence-build-<schema>`) which can have a default fallback to just `docker build --quiet`.

Don't get me wrong I'm not insisting on my solution. I'm just not happy with the `$PATH`-based solution since as for me it's too dirty.

> We're introducing a docker compose integration in a coming release. The compose spec may well contain services that are built on demand. I don't think we can swap the binary when that happens. Do you have a solution for this already?

That's why I was concerned about `$PATH` variant. For me, it was intuitively not flexible enough. I just don't have enough arguments against it from the top of my head.

> it should be extended as a docker cli-plugin (but AFAICT, docker doesn't allow the build command to be extended).

I agree. Last time I checked docker builder's code. Docker supports url-like contexts via the chain of `if-else`-statements and they support only `git...` and `https...` schemas. So if you say use [AWS CodeCommit](https://aws.amazon.com/codecommit/) repo or something else, you can't use it as a docker context. While in git you can:

```bash
git clone codecommit://bla-bla-bla  # works fine
git clone aws+ec2::http://ec2.us-west-2.amazonaws.com/describe-instances  # also possible

docker build codecommit://bla-bla-bla  # will fail
```

> Telepresence has a config.yml file where we can add new structured properties...

This one intuitively sounds good. Need some time to think about it.

Thank you for the detailed response!

**Comment by thallgren** [comment]
> It's not necessary to implement a full stdin-stdout protocol

But it is, and this is perhaps the most important point of all. A generic extension mechanism for `docker build` is not created *only* to cater for your specific use-case. From our side, it's a generic way to extend the Telepresence client, and as such, there's an API to agree on. We will expect a binary that does what the docs for `docker build` stipulate, and we are bound to that expectation as our code evolves. You are too, and you must not assume that we use `docker build` in a specific way. Things will change over time.

---

Questions I'd like an answer to:

- How do you currently handle a docker compose service with a `build` instruction? The build may happen when you do `docker compose up`.
- A [file watcher](https://docs.docker.com/compose/file-watch/) may trigger a build at any time. How do you ensure that your builder is used? Supporting this is a very important part of our docker integration.
- What specific problems did you run into when you tested the dual PATH approach?

Neither the dual PATH workaround, nor my proposed solution to extend `docker build`, will cover the docker compose use-cases.
